### PR TITLE
Configuration file formats for JSON, INI, YAML and TOML

### PR DESCRIPTION
**Pull request by infinisil**
###### Motivation for this change
This is a main part of the implementation of NixOS/rfcs#42 (see https://github.com/Infinisil/rfcs/blob/config-option/rfcs/0042-config-option.md#part-1-1).

Since that part of the RFC is fairly uncontroversial, I think we can merge this before the RFC is eventually accepted. This allows people to use such types in their NixOS options already.

Ping @Profpatsch @rycee @aanderse @kampka 

###### Things done

- [x] Tested that these types and generator combinations indeed can be used in modules.
- [x] Write some tests
- [x] Wrote documentation on how to use them..

Here's a simple example module though:
```nix
{ lib, config, pkgs, ... }:
let
  format = pkgs.formats.json {};
in {
  options.settings = lib.mkOption {
    type = format.type;
    default = {};
  };

  config = {
    settings.foo = 10;
    settings.bar.baz = "Hello";
    environment.etc.test.source = format.generate "config.json" config.settings;
  };
}
```

### CODE REVIEW

**Comment by Infinisil** [comment]
Potentially/Eventually these formats might be implemented using `lazyAttrsOf` (or what becomes of it) from https://github.com/NixOS/nixpkgs/pull/70138, such that self-dependent assignments of them will be possible.

**Comment by aanderse** [comment]
@Infinisil awesome stuff!

File: lib/settings-formats.nix
Comment (Profpatsch) [inline lib/settings-formats.nix]: This is not recursive, is it?

Comment (Infinisil) [reply lib/settings-formats.nix]: It is set on `value`, which is used in the definition itself, but the `description` doesn't refer to the subtypes, so it won't be infinitely long

Comment (Profpatsch) [reply lib/settings-formats.nix]: Ah, missed that.

**Review by Profpatsch — APPROVED** [review]

**Comment by Infinisil** [comment]
(This still needs a bit of docs probably and a history cleanup at least btw, before anybody merges)

**Comment by Infinisil** [comment]
I've now rebased this ontop of https://github.com/NixOS/nixpkgs/pull/70138 to implement types where values can depend on other values. E.g. this previously threw an infinite recursion error, but now it doesn't anymore:

```nix
{ lib, config, ... }:
let format = lib.formats.json;
in {
  options.settings = lib.mkOption {
    type = format.type;
    default = {};
  };

  config = {
    settings.foo = true;
    settings.bar = lib.mkIf config.settings.foo "Foo is true!";

    environment.etc.test.text = format.generate config.settings;
  };
}
```

Due to how the implementation works, it's however a bit more cumbersome if you aren't sure whether an attribute exists at all. This is how that can now be handled gracefully:
```nix
{
  # To make sure the `.name` attribute exists
  settings.name = format.empty;
  # To handle the case of `.name` being empty
  settings.greeting = "Hello, ${format.get "nobody" config.settings.name}!";
}
```

This will generate `{ greeting = "Hello, nobody!"; }`

Alternatively if you don't mind `name` being set:
```nix
{
  settings.name = lib.mkDefault "nobody";
  settings.greeting = "Hello, ${config.settings.name}!";
}
```

Which will generate `{ name = "nobody"; greeting = "Hello, nobody!"; }`

File: lib/settings-formats.nix
Comment (Infinisil) [inline lib/settings-formats.nix]: It might be a good idea to only have a single empty value for all

**Comment by Infinisil** [comment]
On IRC @adisbladis brought to my attention that to support more formats like TOML, we could use `builtins.toJSON` and convert that to toml with a build-time tool, meaning the generated strings won't be available to nix at eval time anymore and the generate functions have the signature `generate :: Value -> Derivation`. This doesn't seem like a problem: Nobody should need the raw Nix strings anyways, and this will speed up evaluation too.

**Comment by abbradar** [comment]
I have a seemingly working `toTOML` which is used for some configuration files: https://gist.github.com/abbradar/f59b55fbe3e80f1b7eab73cf10a4714e . I can check it just in case and PR it to this branch in a few days. cc @pstn

**Comment by Infinisil** [comment]
Nice! Would be best to add it to `lib/generators.nix`

**Comment by Infinisil** [comment]
Though I quite like the idea of doing this at build-time as mentioned above, which will also give more flexibility to change the implementation in the future. So this might not be needed for this PR specifically, but it would still be nice for other use cases in `lib.generators`.

**Comment by Infinisil** [comment]
I now changed it to allow built-time conversions and was able to add a toml type and generator because of this. Also moved the `empty` and `get` functions outside the formats to reduce duplication. A module using this now looks like this:

```nix
{ lib, config, pkgs, ... }:
let
  format = lib.formats.toml;
in {
  options.settings = lib.mkOption {
    type = format.type;
    default = {};
  };

  config = {
    settings.foo.bar = lib.formats.empty;
    settings.bar.foo = "test${lib.formats.get "nothing" config.settings.foo.bar}";
    environment.etc.test.source = format.generate pkgs "config.toml" config.settings;
  };
}
```

**Comment by Infinisil** [comment]
Todo: Add `__toString` to the empty value

**Comment by Infinisil** [comment]
Todo: Change `empty` -> `unset`

**Comment by Infinisil** [comment]
Todo: `get` -> `getOrDefault`

**Comment by Mic92** [comment]
I would also love to see this for home-assistant, where my configuration is getting really large, but I cannot split it into different modules properly.

**Comment by danbst** [comment]
is this ready for merge?

The only thing I wish is `jq .`-ing files before saving them. But I guess this can be added in a separate PR as a generator `lib.formats.jsonPretty`.

**Comment by Infinisil** [comment]
Nah, I'm in the process of writing decent tests for this. Also needs some docs still..

Nice idea though! With the generation potentially happening at build time now, using `jq` to prettify it will be very easy! I'll do that too

File: lib/settings-formats.nix
Comment (Profpatsch) [inline lib/settings-formats.nix]: if you use `pkgs`, this cannot live in `lib/` anymore; anything involving build-time should like in `pkgs/build-support`.

Though maybe it could be split, so that the pure nix parts still live in `lib/`.

File: lib/generators.nix
Comment (Profpatsch) [inline lib/generators.nix]: nope, not a good idea. Look at the bottom of the function, there is already a case for `isFloat`.

File: lib/settings-formats.nix
Comment (Profpatsch) [reply lib/settings-formats.nix]: Edit: saw the `generate` structure. If we need pkgs for these functions, I’m really not sure this is a good fit for `lib/`.

File: lib/tests/formats.nix
Comment (Profpatsch) [inline lib/tests/formats.nix]: yeah, `--read-write-mode` is a good indicator that this depends on import-from-derivation.

**Review by Profpatsch — DISMISSED** [review]
Not an in-depth review, but I have two things:

File: lib/settings-formats.nix
Comment (Infinisil) [reply lib/settings-formats.nix]: Also thought about this, but I'm not entirely sure what to do, because the `type` part doesn't need `pkgs` and it makes for sense for it to be in `lib`, unlike the `generate` part. However those two parts should be very close together, so splitting them in `pkgs` and `lib` isn't optimal either. I guess putting both in `pkgs.formats.{json,ini,yaml,toml}.{type,generate}` might be the best solution, even though I don't really like it.

File: lib/generators.nix
Comment (Infinisil) [reply lib/generators.nix]: Oh, didn't see that. Other than Nix not printing floats with many decimals, what's the problem with this? Having this case supported here will make things a lot easier, and I don't think there's a case where it can cause harm.

File: lib/tests/formats.nix
Comment (Infinisil) [reply lib/tests/formats.nix]: Yeah, I'm in the process of making it not depend on that

File: lib/settings-formats.nix
Comment (danbst) [reply lib/settings-formats.nix]: @Profpatsch I'd rather vote to move `pkgs/build-support/trivial-builders.nix` to `lib`. Let's leave "pure" lib to `<nix/lib>`, not `<nixpkgs/lib>`.

Comment (Infinisil) [reply lib/settings-formats.nix]: @danbst I think a problem with this is that there's no reasonable default `pkgs` to use for library functions that depend on it. The only choice would be to use `import ../. {}`, however then for NixOS, that nixpkgs is imported and evaluated twice, once for `lib`, and once for `pkgs` itself (with potential overlays and stuff).

File: lib/generators.nix
Comment (Profpatsch) [reply lib/generators.nix]: Nix just silently removes decimals, so it’s really not an option to print them.

Comment (Infinisil) [reply lib/generators.nix]: Yeah but I don't think it's a problem. It seems that Nix prints 6 relevant digits, which should be enough for configurations, which is pretty much what Nix is for. I don't see how this could cause any harm.

File: lib/settings-formats.nix
Comment (Infinisil) [reply lib/settings-formats.nix]: The changes in this PR now live under `nixpkgs/pkgs/pkgs-lib`, which is intended for things that would fit into `nixpkgs/lib` but have a dependency on `pkgs`

File: lib/tests/formats.nix
Comment (Infinisil) [reply lib/tests/formats.nix]: Done now

File: lib/generators.nix
Comment (Profpatsch) [reply lib/generators.nix]: Silently removing *anything* is not an option. Big :-1: on this change.

File: pkgs/pkgs-lib/default.nix
Comment (Profpatsch) [inline pkgs/pkgs-lib/default.nix]: There should be a docstring or README.md describing the idea behind pkgs-lib.

File: lib/generators.nix
Comment (Infinisil) [reply lib/generators.nix]: Well I need to show floats in this PR. I guess if you're against it here I'll just have to put it somewhere else where it can be used by the functions.

Comment (Infinisil) [reply lib/generators.nix]: https://github.com/NixOS/nixpkgs/pull/75584/commits/a39fe137ade65c9351740f2d9792674004f995fc

Comment (Infinisil) [reply lib/generators.nix]: As suggested by @emilazy on IRC, it's possible to detect when an inaccurate conversion would be done. Will do that and only then throw an error

Comment (Infinisil) [reply lib/generators.nix]: Did the thing in the latest commits now

**Comment by Infinisil** [comment]
It's probably a good idea to have a generic way to pass parameters to these formats, like how all generators in `lib.generators` take an attribute set with some options you can configure (cc @Profpatsch). E.g. the INI format has varying features across implementations: https://en.wikipedia.org/wiki/INI_file#Varying_features, all of which could be supported with additional parameters to the `lib.generators.toINI` function: https://github.com/NixOS/nixpkgs/blob/169fc2eecaf48348d1d2dc81b500fb55a94518e6/lib/generators.nix#L105-L110 (which I just discussed with @aanderse for https://github.com/NixOS/nixpkgs/pull/81940#discussion_r389391419)

If we don't support such parameters we'll probably end up with multiple different INI types over time, e.g. I know at least one config in NixOS needs [global properties](https://en.wikipedia.org/wiki/INI_file#Global_properties).

I'll take a look at implementing this.

**Comment by Profpatsch** [comment]
> I'll take a look at implementing this.

It might make sense to have multiple types if the differences are too big, too avoid elisp-style kitchen sink functions. e.g. toINI can’t be made much more configurable without losing any sense of structure (it already assumes very little, that headers are `[` and `]` and that you need some way of escaping header contents and rendering key-value pairs).

**Review by danbst — APPROVED** [review]
ready for merge?

**Comment by Infinisil** [comment]
Nope! Change of plan, see https://github.com/NixOS/rfcs/pull/42#issuecomment-599714033 and https://github.com/NixOS/nixpkgs/pull/82743, which allows for a much nicer way to implement the RFC, removing a lot of complexity from this PR.

**Comment by danbst** [comment]
@Infinisil what's the problem to merge this first? RFC can be resolved later.

I think `unstructuredType = pkgs.formats.json.type;` would be fine for typed modules. Is API for formats going to be revamped THAT much?

**Comment by Infinisil** [comment]
@danbst Yeah it will change a bit, specifically all the bits regarding `lazyAttrsOf` and `unset` won't be needed anymore, because with https://github.com/NixOS/nixpkgs/pull/82743, such an unstructured option recursively depending on itself shouldn't be a problem anymore (because you'd explicitly declare the option you depend on). This also has nicer semantics.

**Comment by Infinisil** [comment]
Actually, I think this is needed in its current state after all. While with https://github.com/NixOS/nixpkgs/pull/82743 it's possible to get around infinite recursion by declaring an option, you might not always want to do that. With the `lazyAttrsOf` and `unset` bits here you can avoid infinite recursion even without declaring options for everything that depends on other things.

**Comment by Profpatsch** [comment]
@Infinisil is this PR ready for final review?

**Comment by Infinisil** [comment]
There's still things to do, I'll try to get to them soon. This includes:
- Adding an attrset parameter to the generators to support configuring them
- Making the INI format support the `listsAsDuplicateKeys` parameter (seen this being needed twice now)
- Finishing the docs

**Comment by Infinisil** [comment]
Still not done, but update:
- The whole `unset` and `lazyAttrsOf` business is just damn ugly and brittle, I couldn't bring myself to leave this in. With https://github.com/NixOS/nixpkgs/pull/82743, this really won't be necessary anymore for recursive assignments.
- Formats now support passing of parameters, the `listsAsDuplicateKeys` thing is implemented for INI

Still need to finish the docs though

**Comment by Mic92** [comment]
We have now several copies for yaml/json types laying around in different NixOS modules. Should I just factor out the yaml type and make a new PR?

**Comment by Infinisil** [comment]
Took some time to finish writing the docs for this. Would love somebody to proof-read or tell me how they can be improved (cc @domenkozar since I know you're pretty good at writing docs!):

Screen Shot 2020-07-25 at 21 47 03

File: nixos/doc/manual/development/settings-options.xml
Comment (Mic92) [inline nixos/doc/manual/development/settings-options.xml:13]: It's really annoying that our xml formatter does not break lines.
This is very hard to read on github.

Comment (Infinisil) [reply nixos/doc/manual/development/settings-options.xml:13]: Oh good point. I could introduce manual line breaks, but those get annoying quickly when the text needs changes.

Comment (jtojnar) [reply nixos/doc/manual/development/settings-options.xml:13]: This is [intentional](https://github.com/NixOS/nixpkgs/pull/68181) since hard wrapping breaks diffs. On the other hand, it is easy to enable soft wrapping using an user style. I recommend https://github.com/StylishThemes/GitHub-code-wrap.

File: nixos/doc/manual/development/settings-options.xml
Comment (jtojnar) [inline nixos/doc/manual/development/settings-options.xml:160]: Maybe we should avoid showing adding port to firewall whitelist since modules are not supposed to do that.

File: nixos/doc/manual/development/settings-options.xml
Comment (Mic92) [reply nixos/doc/manual/development/settings-options.xml:13]: I see. The only alternative that would make both github code review and diffs happy is to use
break on every sentence instead of each paragraph. No idea if our xmlformatter even would have an option for that.

File: nixos/doc/manual/development/settings-options.xml
Comment (Infinisil) [reply nixos/doc/manual/development/settings-options.xml]: Oh good idea! I changed it to a `user` option now.

File: pkgs/pkgs-lib/default.nix
Comment (Infinisil) [reply pkgs/pkgs-lib/default.nix:4]: @Profpatsch Just added a docstring explaining pkgs-lib: https://github.com/NixOS/nixpkgs/pull/75584/files#diff-fa126a2182114f76a8e8df7ed256395d. Does that look good?

File: nixos/doc/manual/development/settings-options.xml
Comment (domenkozar) [reply nixos/doc/manual/development/settings-options.xml:13]: Given that we're months away switching to readable documentation I wouldn't bother right now.

File: nixos/doc/manual/development/settings-options.xml
Comment (mdlayher) [inline nixos/doc/manual/development/settings-options.xml:13]: Super nit: is Oxford comma (or lack of) a convention in Nix docs: e.g. "INI, YAML, and TOML."?

**Review by mdlayher — APPROVED** [review]
LGTM! I think the docs you've provided and example module are helpful. I'd be happy to be an early adopter of this package since my CoreRAD module already uses the settings attribute with a bespoke Nix -> JSON -> TOML conversion: https://github.com/NixOS/nixpkgs/pull/89781.

**Review by Profpatsch — COMMENTED** [review]
Haven’t reviewed in-depth, but if other people are okay with it LGTM

File: nixos/doc/manual/development/settings-options.xml
Comment (Infinisil) [reply nixos/doc/manual/development/settings-options.xml:13]: I didn't even know there could be a comma there! There isn't a convention I'm fairly certain, so I'll just leave it as is for now :)

**Comment by Infinisil** [comment]
I think this PR is in a good state now. Unless somebody has any improvement suggestions or concerns I'll merge it this weekend!

**Comment by domenkozar** [comment]
I'd review it if it was written with terminal length in mind, but currently it's not feasible via github ui.

File: pkgs/pkgs-lib/default.nix
Comment (roberth) [inline pkgs/pkgs-lib/default.nix:5]:
```suggestion
  # setting format types and generators. These do not fit in lib/types.nix, because
```

File: pkgs/pkgs-lib/default.nix
Comment (roberth) [inline pkgs/pkgs-lib/default.nix:1]:
```suggestion
# pkgs-lib is for functions that can't be in lib, because
```

File: pkgs/top-level/stage.nix
Comment (roberth) [inline pkgs/top-level/stage.nix:78]: An overlay is unnecessary because the same could be achieved by putting `formats` in `all-packages.nix`.

Categorizing functions separately from packages is beyond the scope of this change. It can be discussed in a separate issue/pr, although I doubt its usefulness.

**Review by roberth — CHANGES_REQUESTED** [review]
It should be noted in the docs that some of these functions may write to the store, which is not safe when Nix is handling secrets. This is in contrast to `builtins.toJSON` etc, which operate entirely in memory.

**Comment by worldofpeace** [comment]
> I'd review it if it was written with terminal length in mind, but currently it's not feasible via github ui.

this is useful https://github.com/StylishThemes/GitHub-code-wrap

**Comment by rycee** [comment]
An alternative that doesn't need any installations is to use the "Split" diff option. For whatever reason, it seems to wrap lines.
github-split

**Comment by Infinisil** [comment]
> It should be noted in the docs that some of these functions may write to the store, which is not safe when Nix is handling secrets. This is in contrast to `builtins.toJSON` etc, which operate entirely in memory.

For NixOS modules at least it doesn't make much of a difference, because the only way for a string (which `toJSON` returns) to influence a system build is by putting it in a derivation. With the interface this PR adds, this step is just done a little earlier. But yeah it can't hurt to point this out, I'll add a note

File: pkgs/top-level/stage.nix
Comment (Infinisil) [reply pkgs/top-level/stage.nix:78]: Doesn't have to be just functions though, can be arbitrary values, and other functions like `fetchgit` wouldn't fit in a pkgs-lib. I guess it won't be a problem to just put `formats` in `all-packages.nix`, but imo having a separation between "things needed to create packages" and "things that just happen to depend on pkgs" would make sense. Could still introduce that in the future though

File: lib/generators.nix
Comment (Infinisil) [reply lib/generators.nix:52]: Done

File: pkgs/pkgs-lib/default.nix
Comment (Infinisil) [reply pkgs/pkgs-lib/default.nix]: Done with https://github.com/NixOS/nixpkgs/compare/76e8e4a26fa451bbc1832e0cd599080795227593..83b16885f526d6ab7b39e98159e2b48024f3238c#diff-fa126a2182114f76a8e8df7ed256395d

File: pkgs/pkgs-lib/default.nix
Comment (Infinisil) [reply pkgs/pkgs-lib/default.nix]: Done with adjustments in https://github.com/NixOS/nixpkgs/compare/76e8e4a26fa451bbc1832e0cd599080795227593..83b16885f526d6ab7b39e98159e2b48024f3238c#diff-fa126a2182114f76a8e8df7ed256395d

File: pkgs/top-level/stage.nix
Comment (Infinisil) [reply pkgs/top-level/stage.nix]: Not using an overlay anymore now

Comment (roberth) [reply pkgs/top-level/stage.nix]: All of `formats`, `fetchgit`, `dockerTools` depend on Nixpkgs, take arguments, produce derivations, are maintained and released in the same way. What makes `formats` so architecturally different that it needs to be treated differently?

> imo having a separation between "things needed to create packages" and "things that just happen to depend on pkgs" would make sense. Could still introduce that in the future though

If only one attribute is separate it doesn't make sense though. For it to make sense it must be applied to the whole of Nixpkgs and it must have practical benefits. The prior can be done in the future and I'm skeptical about the benefits, but would love to be proven wrong in that same future.

For now please follow the example of, say, `dockerTools`.

Comment (roberth) [reply pkgs/top-level/stage.nix]: Looks like my view of the PR was outdated (maybe caused by the current service disruption?) anyway, thanks for removing the overlay. Currently `build-support` has the role of `pkgs-lib` so you may want to move the files to `build-support/formats`. I don't really like that name either so I'm going to leave this change up to you to decide.

Comment (Infinisil) [reply pkgs/top-level/stage.nix]: The difference between e.g. `fetchgit` and `formats.json` is that `fetchgit` is needed for building many of the `pkgs.*` derivations, whereas `formats.json` doesn't have any use for any `pkgs.*`. Instead `formats` are mostly just useful for NixOS modules (really, the module system in general, so it wouldn't fit in `nixos/lib`)

This is also why I don't think `build-support` would make sense, because `formats.json` doesn't support any package building.

**Comment by Infinisil** [comment]
@roberth Looking good now?

**Review by domenkozar — APPROVED** [review]

**Review by roberth — APPROVED** [review]

**Comment by Ericson2314** [comment]
I wonder if `pkgs-lib` should be `pkgs/lib` to match `nixos/lib`.

**Comment by Infinisil** [comment]
Awesome, thanks for merging!

@Ericson2314 I intentionally didn't name it `lib` because it could easily be confused with `pkgs.lib`.

**Comment by fzakaria** [comment]
@Infinisil you might want to update this documentation section as well https://nixos.org/manual/nixpkgs/unstable/#sec-generators
It seems out of date considering the new NixOS doc you added.

**Comment by fzakaria** [comment]
Actually reading `lib/generators.nix` it looks like it should be consolidated with the new **formats** option introduced here.

**Pull request merged**
